src/SigninRequest.ts

```diff
@@ -94,6 +95,38 @@ export class SigninRequest {
             client_secret, scope, extraTokenParams,
             skipUserInfo,
         });
+    }
+
+    public async getUrl(): Promise<string> {
+        const nonOptionalParams: Array<keyof SigninRequestArgs> = [
```

Comment:
this right here is what i meant.

we can't rely on object rest to get `...optionalParams` because it'd contain properties previously not included in that set (the unused vars). 

alternatively, we could flip this around and have an array of optional prop names.

whichever way we do it, we need an array of keys i think. 

the way i have it now, if we pass unknown properties (type errors) at runtime as a js consumer, they'll make it through to the query string.

but i admit whatever we do its not great looking

Reply:
You are right both way are not really nice. This does it more explicit. Also more code... I like this version, but can live with the first version you had with the eslint ignore unused variables...
```js
export class SigninRequest {
    private readonly _logger = new Logger("SigninRequest");

    public readonly state: SigninState;

    private readonly _url: string;
    private readonly _client_id: string;
    private readonly _redirect_uri: string;
    private readonly _response_type: string;
    private readonly _scope: string;
    private readonly _response_mode?: "query" | "fragment";
    private readonly _nonce?: string;
    private readonly _resource?: string | string[];
    private readonly _extraParams: Partial<SigninRequestArgs>;

    public constructor({
        // mandatory
        url, authority, client_id, redirect_uri, response_type, scope,
        // optional
        state_data, response_mode, request_type, client_secret, nonce,
        resource,
        skipUserInfo,
        extraQueryParams,
        extraTokenParams,
        disablePKCE,
        ...extraParams
    }: SigninRequestArgs) {
        if (!url) {
            this._logger.error("ctor: No url passed");
            throw new Error("url");
        }
        if (!client_id) {
            this._logger.error("ctor: No client_id passed");
            throw new Error("client_id");
        }
        if (!redirect_uri) {
            this._logger.error("ctor: No redirect_uri passed");
            throw new Error("redirect_uri");
        }
        if (!response_type) {
            this._logger.error("ctor: No response_type passed");
            throw new Error("response_type");
        }
        if (!scope) {
            this._logger.error("ctor: No scope passed");
            throw new Error("scope");
        }
        if (!authority) {
            this._logger.error("ctor: No authority passed");
            throw new Error("authority");
        }

        this._url = url;
        this._client_id = client_id;
        this._redirect_uri = redirect_uri;
        this._response_type = response_type;
        this._scope = scope;
        this._response_mode = response_mode;
        this._client_id = client_id;
        this._nonce = nonce;
        this._resource = resource;
        this._extraParams = extraParams;

        this.state = new SigninState({
            data: state_data,
            request_type,
            code_verifier: !disablePKCE,
            client_id, authority, redirect_uri,
            response_mode,
            client_secret, scope, extraTokenParams,
            skipUserInfo,
        });
    }

    public async getUrl(): Promise<string> {
        const parsedUrl = new URL(this._url);
        parsedUrl.searchParams.append("client_id", this._client_id);
        parsedUrl.searchParams.append("redirect_uri", this._redirect_uri);
        parsedUrl.searchParams.append("response_type", this._response_type);
        parsedUrl.searchParams.append("scope", this._scope);
        if (this._nonce) {
            parsedUrl.searchParams.append("nonce", this._nonce);
        }

        parsedUrl.searchParams.append("state", this.state.id);

        const challenge = await this.state.getChallenge();
        if (challenge) {
            parsedUrl.searchParams.append("code_challenge", challenge);
            parsedUrl.searchParams.append("code_challenge_method", "S256");
        }

        if (this._resource) {
            // https://datatracker.ietf.org/doc/html/rfc8707
            const resources = Array.isArray(this._resource) ? this._resource : [this._resource];
            resources
                .forEach(r => parsedUrl.searchParams.append("resource", r));
        }

        for (const [key, value] of Object.entries({ response_mode: this._response_mode, ...this._extraParams })) {
            if (value != null) {
                parsedUrl.searchParams.append(key, value.toString());
            }
        }

        return parsedUrl.href;
    }
}
```

Reply:
thats true, just putting them on `this` individually. i think i'd prefer that over relying on unused vars as it seems they could fall over/be a bit flaky over time.

ill give it a try